File: fun/odd-even.c
#include <stdio.h>
#include <time.h>

// NOTE: can't be bothered to learn how to do this with minimal code duplication in C for now

/*
Results:
  These tests were so much faster than python I had add two more orders of magnitude
  to the test lengths...

  I kept seeing almost no differences until I added this global variable. I believe
  the compiler was just removing the line of code before that since it was deemed "useless"
*/

int result;

double runConventionalTests(int numTests)
{
  clock_t t;
  t = clock();
  for (int i = 0; i < numTests; i++)
  {
    result = i % 2;
  }
  t = clock() - t;
  return ((double)t) / CLOCKS_PER_SEC;
}

double runBitwiseTests(int numTests)
{
  clock_t t;
  t = clock();
  for (int i = 0; i < numTests; i++)
  {
    result = i & 1;
  }
  t = clock() - t;
  return ((double)t) / CLOCKS_PER_SEC;
}

int main()
{
  int iterations[7] = {1000, 10000, 100000, 1000000, 10000000, 100000000, 1000000000};

  for (int i = 0; i < 7; i++)
  {

    int numIterations = iterations[i];

    double conventionalTestTime = runConventionalTests(numIterations);
    printf("conventionalTestTime() took %f seconds to execute %d iterations\n", conventionalTestTime, numIterations);

    double bitwiseTestTime = runBitwiseTests(numIterations);
    printf("bitwiseTestTime() took %f seconds to execute %d iterations\n", bitwiseTestTime, numIterations);
  }
  return 0;
}

File: fun/odd-even.py
# How much faster is odd/even through bit-shifting faster than a conventional approach?

"""
Results:
	interesting because bitwise is actually slightly slower... but that may just be python
"""

import time


def conventional_num_is_odd(num):
	return num % 2


def bitwise_num_is_odd(num):
	return num & 1


# make it more difficult for any optimizations to occur
some_global_var = 0


def _perform_test(fn, num_iterations):
	start = time.time()
	for i in range(num_iterations):
		global some_global_var
		some_global_var = fn(i)
	return time.time() - start


for num_iterations in [1000, 10000, 100000, 1000000, 10000000]:
	print(f"Testing with {num_iterations} iterations")

	print("----> testing conventional")
	conventional_time = _perform_test(conventional_num_is_odd, num_iterations)
	print(f"----> conventional took {conventional_time} seconds")

	print("----> testing bitwise")
	conventional_time = _perform_test(bitwise_num_is_odd, num_iterations)
	print(f"----> bitwise took {conventional_time} seconds")

